src/Features/Core/Portable/MetadataAsSource/DecompilationMetadataAsSourceFileProvider.cs

```diff
@@ -94,20 +82,25 @@ internal sealed class DecompilationMetadataAsSourceFileProvider(IImplementationA
         var infoKey = await GetUniqueDocumentKeyAsync(sourceProject, topLevelNamedType, signaturesOnly: !useDecompiler, cancellationToken).ConfigureAwait(false);
 
         var fileInfo = _keyToInformation.GetOrAdd(infoKey,
-            _ => new MetadataAsSourceGeneratedFileInfo(tempPath, sourceWorkspace, sourceProject, topLevelNamedType, signaturesOnly: !useDecompiler));
+            _ => new MetadataAsSourceGeneratedFileInfo(sourceWorkspace, sourceProject, topLevelNamedType, signaturesOnly: !useDecompiler, persister));
 
-        DocumentId generatedDocumentId;
+        var generatedDocumentId = fileInfo.DocumentId;
         Location navigateLocation;
-        if (!_generatedFilenameToInformation.TryGetValue(fileInfo.TemporaryFilePath, out var existingDocumentId))
+
+        var existingDocument = metadataWorkspace.CurrentSolution.GetDocument(fileInfo.DocumentId);
```

Comment:
removed the `_generatedFilenameToInformation` as that can be tracked by the workspace (since we're storing the document in there anyway, that was feedback from the previous PR)